Make type optional

compwright opened the pull request:
Resolves #41

compwright committed:
Don’t require `type` or `typeField` if `elasticsearch.apiVersion` >= 6

Resolves #41

AntonBazhal commented on README.md:
```diff
@@ -22,6 +22,7 @@ $ npm install dynamo2es-lambda
 - **type** - { String } - Elasticsearch type to be used for all the documents; optional if `typeField` is provided
 - **[elasticsearch - alias: es]** - { Object } - Elasticsearch configuration; under the hood library uses [aws-elasticsearch-client][aws-elasticsearch-client-url]; for more information check [this documentation][aws-elasticsearch-client-url]
   - **[bulk]** - { Object } - aside from general Elasticsearch configuration, you can use this field to pass additional parameters to [bulk API][bulk-api-url]
+- **[elasticsearchVersion]** - { Number } - if >= 6, disables the requirement for a `type` or `typeField` option (mapping types are [deprecated in ES 6](https://www.elastic.co/guide/en/elasticsearch/reference/6.0/removal-of-types.html))
```
`elasticsearch` config option takes an optional `apiVersion` parameter, so I'd prefer to use it instead of adding a new one. https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/16.x/configuration.html

compwright replied:
Yeah, that makes total sense. I've assumed the default to be `5` to preserve the current default behavior, but the current Elasticsearch default is `7` so this should probably be bumped and the tests updated accordingly in the next major version of dynamo2es-lambda.

AntonBazhal commented on lib/schemas.js:
```diff
@@ -2,61 +2,67 @@
+
+  const FIELD = joi.string().min(1);
+
+  const ELASTICSEARCH_SCHEMA = joi.object({
```
This schema can be extended to have an optional `apiVersion` parameter with a default set to `'5'`

AntonBazhal replied:
Or even better: don't set a default and treat missing as v5 (type required)

AntonBazhal commented on lib/handler.js:
```diff
@@ -16,10 +16,11 @@ const DEFAULT_RETRY_COUNT = 0;
 
   const elasticsearchOpts = options.elasticsearch || options.es || {};
   const bulkOpts = elasticsearchOpts.bulk || {};
+  const apiVersion = Number(elasticsearchOpts.apiVersion) || 5;
```
This can be a part of the schema. See my comment below.

AntonBazhal commented on lib/handler.js:
```diff
@@ -66,6 +67,10 @@ module.exports = function(options) {
                 _id: id
               };
 
+              if (apiVersion >= 6 && !actionDescriptionObj._type) {
```
According to the docs, `apiVersion` is a string and not just a major version. Thus, it makes sense to use `semver` to compare versions.

AntonBazhal commented on lib/schemas.js:
```diff
@@ -2,61 +2,67 @@
+    .oxor('idField', 'idResolver')
+    .oxor('versionField', 'versionResolver')
+    .xor('index', 'indexField')
+    [apiVersion >= 6 ? 'oxor' : 'xor']('type', 'typeField') // eslint-disable-line no-unexpected-multiline
```
I think this can be based on schema using `when` syntax

compwright commented:
The Joi `.when()` syntax proved to be pretty tricky to master, but this should do it.

AntonBazhal commented on lib/handler.js:
```diff
@@ -60,10 +60,22 @@ module.exports = function(options) {
                 _id: id
               };
 
+              // Omit blank _type when ES >= 6
```
The new ES client library does not support `apiVersion` attribute. Since there is a plan to upgrade this library to use it, should we just allow the type to be not present without checking the version? Thoughts?

compwright replied:
I think that would be cleaner. I was trying not to break backward compatibility, but if you're willing to make a minor or major release then I see no problems and it would be more future-proof.

compwright replied:
@AntonBazhal FYI the reason I did it this way is because of what you said here: https://github.com/AntonBazhal/dynamo2es-lambda/issues/41#issuecomment-388831433

If you decide you still want the type config to be validated, just delete the last commit on this branch. Or, perhaps you could do both major and minor releases if you want to support older clients.

compwright committed:
Bump required node version

compwright committed:
Make type config optional in all versions

Breaks backward-compatibility, but allows for better future-proofing, and simpler code. We’ll just rely on ES itself to throw an error if `_type` is required.

AntonBazhal commented on lib/schemas.js:
```diff
@@ -1,11 +1,12 @@
     body: joi.any().forbidden()
-  }).optional().unknown()
+  }).optional().unknown(),
+  apiVersion: joi.string().default('5')
```
There is no need for this config any more

compwright replied:
Ah yes, missed that one. Thanks.

AntonBazhal reviewed (commented):
One minor comment and should be good to merge
